### ISSUE TITLE
Support authentication of nodes

### ISSUE THREAD

**Issue by alexlarsson** [engelmi/hirte#88]
We should be able to define in the manager configuration details about each node, and this should include authentication information for each of the nodes (like a client certification or similar) such that we can verify that the right node is connecting.

**Comment by pypingou** [comment]
Would this be covered by Ygal's work on using the TLS/SSL proxy and their certificates?

**Comment by mwperina** [comment]
I don't think so, TLS encryption using HA proxy just covers agent-manager connection encryption

**Comment by pypingou** [comment]
> I don't think so, TLS encryption using HA proxy just covers agent-manager connection encryption

Depends, if you have certificates with a shared CA, that can also be used for
authentication

**Comment by dougsland** [comment]
if this is not what Ygal has being working and nobody will work. I would like to add into my queue.

**Comment by dougsland** [comment]
Let me investigate this one.

**Comment by pypingou** [comment]
>    Let me investigate this one.

Feel free, but I'm not sure it is/should be one of our priorities

**Comment by dougsland** [comment]
> Let me investigate this one.
> Feel free, but I'm not sure it is/should be one of our priorities

Sure, feel free to redirect me but TLS seems good one to have.

**Comment by ygalblum** [comment]
## TLDR
The answer is as always complicated. The double proxy solution (in my demo HAProxy, but can be other implementation) covers authentication but not authorization. Therefore, it blocks any party without a certificate, but once you have one, nothing stops you from impersonating to any other agent.

## What's the problem
When using mutual TLS (mTLS), both the server and the client present their certificate to allow the other party to authenticate them. For simplicity, let's say they share the same CA to both sign their certificate and authenticate the other party's. With this in place, only parties that have access to a certificate signed by this CA can connect to the server. 

However, this also means that when different cars share the same CA, if they can connect somehow, nothing in this mechanism stops an agent from car A from connecting to the manager in car B. For this purpose, Hirte has the `AllowedNodeNames`. So, assuming that nodes in different cars will have names unique to the car, this covers the innocent cases.

But, what about malicious cases? What stops this node from car A from impersonating to a node from car B? Currently, even with the double proxy, nothing. The reason is that the node name is part of the payload. So, nothing stops that malicious node from setting the node name in the payload to that of the node from car B. Please note that the same goes for a rouge node in car B.

## Solutions
### Partial Solution
I can think of two solutions that can be used reduce the risk and limit it into the nodes inside the car
1. Use a different CAs in each car. 
2. Use the AllowedNodeNames mechanism also in the server's proxy to allow access based on the subject.

In both of this solutions, nodes from other cars will not be able to connect to the server, either because their certificate has a different CA (1) of because the subject of their certificate is not in the allowed list (2). But, nothing is blocking the rouge node in the same car.

### Golden but difficult to implement
The best solution is to get the node name from the certificate's subject and not from the payload. This way you block impersonation altogether since the node's name comes from its certificate. You can see this [example](https://www.haproxy.com/blog/ssl-client-certificate-information-in-http-headers-and-logs) on how details from the SSL certificate may be passed to the backend server in HTTP headers. Having said that, while this example is useful for learning how to extract the information, injecting it into the payload (and in our case the D-Bus message) is more complicated.

**Comment by dougsland** [comment]
Talking with Ygal today, seems we are fine what we have and there are even a layer of security in the chip level (for our main use case). I would just move forward and close this one. Whatever different case show up we can revive this thread. Moving this assign to @ygalblum

**Comment by engelmi** [comment]
You are right @dougsland. Moved the authorization part to a new issue (#364). Closing this one.
